reps/2024-06-13-export-api.md

```diff
@@ -0,0 +1,141 @@
+However, these sources don't need to concern themselves with the details of how events are stored, transmitted, or published.
+
+### Events to Export and How
+#### Events to Export
+- Core
+Tasks/Actors/Objects/Jobs/Nodes/Placement groups, including its meta data and current status when event emits.
+
+|  Event            | Event Source          | When to export                                                                    | Format example(maybe a file link)|
+|  ----             | ----                  | ----                                                                              | ----                             |
+| Tasks             | CoreWorker and raylet | When coreworker adds task event to<br>ray::core::worker::TaskEventBuffer          |          |
+| Actors            | GCS                   | GcsPublisher::PublishActor<br>publish actor status through<br>GCS_ACTOR_CHANNEL   |          |
+| Objects           | CoreWorker and raylet | None                                                                              |          |
+| Jobs              | JobManager            | When JobInfoStorageClient.put_status called                                       |          |
+| Nodes             | GCS                   | GcsNodeManager::HandleXXXNode                                                     |          |
+| Placement groups  | GCS                   | GcsPlacementGroupManager::HandleXXXPlacementGroup                                 |          |
+
+
+- RayServe
+State change events for replicas, deployments, applications.
+
+|  Event            | Event Source          | When to export | Format example(maybe a file link)|
+|  ----             | ----                  | ----           | ----                               |
+| Serve App         | ServeController Actor | None           | None         |
+
+- RayData
+All datasets, the dag, and execution progress.
+
+|  Event            | Event Source          | When to export | Format example(maybe a file link)|
+|  ----             | ----                  | ----           | ----                               |
+| Datasets         | ray.data.internal.stats._StatsActor | _StatsActor's update function called.           | None         |
+
+
+#### How to export
+- Generate event
+We propose to use the filesystem based export interface. The event source can just write the
+event into the file of this node and then return. The export of the event can be the responsibility of the agent on the node. We know that there will be multiple worker processes running multiple actors or tasks on each Ray node. Events generated by multiple worker processes are written to multiple event files, not just one, in order to avoid competition among multiple worker processes or possible file writing errors.
+
+
+- Export event
+We propose to use pubsub for event export, because pubsub is easier to decouple systems,
```

Comment:
My opinion is we can limit this REP to generating the events and let users implement exporting the events to an external system. We can share recommendations on how to do this (eg: log ingestion using Vector which can send data to various sinks per the use case), but I think the pub sub event export isn’t necessary for this version to keep the interface simple and flexible

Reply:
Or instead, we give some possible export methods (such as gcs or vector) as a suggestion, not a proposal?